File: src/main/java/com/example/dpop/orchestrator/api/v1/identification/fsc/FscIdentificationService.java
package com.example.dpop.orchestrator.api.v1.identification.fsc;

import com.example.dpop.account.AccountProfile;
import com.example.dpop.account.AccountService;
import com.example.dpop.ext_stammdaten.ExtStammdatenService;
import com.example.dpop.id_fsc.IdFscService;
import com.example.dpop.orchestrator.account.AccountBindingKeyMappingService;
import com.example.dpop.orchestrator.api.v1.OrchestratorException;
import com.example.dpop.orchestrator.api.v1.OrchestratorResponse;
import com.example.dpop.orchestrator.session.AttemptStatus;
import com.example.dpop.orchestrator.session.ChannelSession;
import com.example.dpop.orchestrator.session.IdentificationAttempt;
import com.example.dpop.orchestrator.session.OrchestratorAttempt;
import com.example.dpop.orchestrator.session.ProcessPurpose;
import com.example.dpop.orchestrator.session.ProcessSession;
import com.example.dpop.orchestrator.session.RegistrationProcessSession;
import com.example.dpop.orchestrator.session.SessionManagementService;
import tools.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Service
@Transactional
public class FscIdentificationService {

    private static final tools.jackson.core.type.TypeReference<Map<String, Object>> MAP_TYPE =
            new tools.jackson.core.type.TypeReference<>() {};

    private final SessionManagementService sessionManagementService;
    private final ExtStammdatenService extStammdatenService;
    private final IdFscService idFscService;
    private final AccountService accountService;
    private final AccountBindingKeyMappingService accountBindingKeyMappingService;
    private final ObjectMapper objectMapper;

    public FscIdentificationService(
            SessionManagementService sessionManagementService,
            ExtStammdatenService extStammdatenService,
            IdFscService idFscService,
            AccountService accountService,
            AccountBindingKeyMappingService accountBindingKeyMappingService,
            ObjectMapper objectMapper
    ) {
        this.sessionManagementService = sessionManagementService;
        this.extStammdatenService = extStammdatenService;
        this.idFscService = idFscService;
        this.accountService = accountService;
        this.accountBindingKeyMappingService = accountBindingKeyMappingService;
        this.objectMapper = objectMapper;
    }

    public OrchestratorResponse startIdentification(UUID channelSessionId, String bindingKeyRef, Map<String, Object> data) {
        ChannelSession channelSession = sessionManagementService.getChannelSessionByBindingKeyRef(bindingKeyRef)
                .orElseThrow(() -> OrchestratorException.forbidden("Channel session not found"));
        ensureChannelMatches(channelSessionId, channelSession);

        RegistrationProcessSession processSession = (RegistrationProcessSession) sessionManagementService
                .getLatestProcessSessionByChannel(channelSession.getChannelSessionId(), ProcessPurpose.REGISTRATION)
                .orElseThrow(() -> new IllegalArgumentException("Process session not found"));

        processSession.setSelectedIdentificationMethod("fsc");
        sessionManagementService.updateProcessSession(processSession);

        IdentificationAttempt attempt = sessionManagementService.createIdentificationAttempt(
                processSession.getProcessSessionId(), "fsc", "input", Duration.ofMinutes(10));

        Map<String, Object> pending = data != null ? new HashMap<>(data) : new HashMap<>();
        List<String> missingFields = computeMissingFscFields(pending);

        if (missingFields.isEmpty()) {
            return verifyFsc(attempt, pending, bindingKeyRef, channelSession, processSession);
        }

        attempt.setStatus(AttemptStatus.INPUT_REQUIRED);
        attempt.setMissingFields(serializeList(missingFields));
        attempt.setResult(toPendingJson(pending));
        sessionManagementService.updateAttempt(attempt);

        return new OrchestratorResponse(
                channelSession.getChannelSessionId(),
                new OrchestratorResponse.ProcessState("REGISTRATION", "ACTIVE", null, null),
                new OrchestratorResponse.AttemptState(attempt.getAttemptId(), "identification", "INPUT_REQUIRED", missingFields, null),
                new OrchestratorResponse.NextRouting("fsc", "input")
        );
    }

    private void ensureChannelMatches(UUID expectedChannelSessionId, ChannelSession actualChannelSession) {
        if (expectedChannelSessionId == null) return;
        if (!actualChannelSession.getChannelSessionId().equals(expectedChannelSessionId)) {
            throw OrchestratorException.forbidden("Channel session mismatch");
        }
    }

    public OrchestratorResponse submitIdentificationData(UUID attemptId, String bindingKeyRef, Map<String, Object> patchData) {
        OrchestratorAttempt attempt = sessionManagementService.getAttemptById(attemptId)
                .orElseThrow(() -> OrchestratorException.notFound("Attempt not found"));

        if (attempt.getStatus() == AttemptStatus.VERIFIED) {
            throw OrchestratorException.conflict("Attempt is already verified");
        }

        Map<String, Object> pending = loadPendingData(attempt);
        if (patchData != null) pending.putAll(patchData);

        List<String> missingFields = computeMissingFscFields(pending);

        if (!missingFields.isEmpty()) {
            attempt.setStatus(AttemptStatus.INPUT_REQUIRED);
            attempt.setMissingFields(serializeList(missingFields));
            attempt.setResult(toPendingJson(pending));
            sessionManagementService.updateAttempt(attempt);
            return new OrchestratorResponse(null, null,
                    new OrchestratorResponse.AttemptState(attemptId, "identification", "INPUT_REQUIRED", missingFields, null),
                    new OrchestratorResponse.NextRouting("fsc", "input"));
        }

        ChannelSession channelSession = sessionManagementService.getChannelSessionByBindingKeyRef(bindingKeyRef)
                .orElseThrow(() -> OrchestratorException.forbidden("Channel session not found"));
        ProcessSession processSession = sessionManagementService.getProcessSessionById(attempt.getProcessSessionId())
                .orElseThrow(() -> new IllegalArgumentException("Process session not found"));

        return verifyFsc(attempt, pending, bindingKeyRef, channelSession, processSession);
    }

    private OrchestratorResponse verifyFsc(OrchestratorAttempt attempt, Map<String, Object> data,
            String bindingKeyRef, ChannelSession channelSession, ProcessSession processSession) {
        UUID attemptId = attempt.getAttemptId();
        String kvnr = (String) data.get("kvnr");
        String fsc = (String) data.get("fsc");

        Long personId = extStammdatenService.findPersonIdByKvnr(kvnr).orElse(null);
        if (personId == null || !idFscService.validateFsc(personId, fsc)) {
            attempt.setStatus(AttemptStatus.FAILED);
            attempt.setResult("{}");
            sessionManagementService.updateAttempt(attempt);
            throw OrchestratorException.verificationFailed("FSC validation failed");
        }

        attempt.setStatus(AttemptStatus.VERIFIED);
        attempt.setResult("{ \"identified\": true, \"personId\": " + personId + " }");
        sessionManagementService.updateAttempt(attempt);

        processSession.setPersonId(personId);
        sessionManagementService.updateProcessSession(processSession);

        AccountProfile account = accountService.identifyAccount(
                personId, "fsc", "HIGH",
                processSession.getProcessSessionId(),
                Map.of("kvnr", kvnr)
        );
        sessionManagementService.setAccountId(channelSession.getChannelSessionId(), account.accountId());
        accountBindingKeyMappingService.mapBindingKeyToAccount(bindingKeyRef, account.accountId());

        boolean hasAuth = !account.activeAuthenticationMethods().isEmpty();
        if (hasAuth) {
            return new OrchestratorResponse(
                    channelSession.getChannelSessionId(),
                    new OrchestratorResponse.ProcessState("REGISTRATION", "ACTIVE", personId, account.accountId()),
                    new OrchestratorResponse.AttemptState(attemptId, "identification", "VERIFIED", null,
                            Map.of("identified", true, "personId", personId)),
                    new OrchestratorResponse.NextRouting("authentication", "selectMethod",
                            account.activeAuthenticationMethods(), null, account.accountId(), personId)
            );
        }
        return new OrchestratorResponse(
                channelSession.getChannelSessionId(),
                new OrchestratorResponse.ProcessState("REGISTRATION", "ACTIVE", personId, account.accountId()),
                new OrchestratorResponse.AttemptState(attemptId, "identification", "VERIFIED", null,
                        Map.of("identified", true, "personId", personId)),
                new OrchestratorResponse.NextRouting("enrollment", "selectMethod", Arrays.asList("sms"),
                        null, account.accountId(), personId)
        );
    }

    public OrchestratorResponse getIdentificationStatus(UUID attemptId, String bindingKeyRef) {
        OrchestratorAttempt attempt = sessionManagementService.getAttemptById(attemptId)
                .orElseThrow(() -> OrchestratorException.notFound("Attempt not found"));

        Object result = null;
        if (attempt.getStatus() == AttemptStatus.VERIFIED && attempt.getResult() != null) {
            result = attempt.getResult();
        }

        return new OrchestratorResponse(null, null,
                new OrchestratorResponse.AttemptState(
                        attempt.getAttemptId(), "identification", attempt.getStatus().toString(), null, result),
                new OrchestratorResponse.NextRouting(attempt.getNextContext(), attempt.getNextStep())
        );
    }

    private List<String> computeMissingFscFields(Map<String, Object> data) {
        List<String> missing = new ArrayList<>();
        if (!data.containsKey("kvnr") || data.get("kvnr") == null) missing.add("kvnr");
        if (!data.containsKey("fsc") || data.get("fsc") == null) missing.add("fsc");
        return missing;
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> loadPendingData(OrchestratorAttempt attempt) {
        String result = attempt.getResult();
        if (result == null || result.isBlank()) return new HashMap<>();
        try {
            Map<String, Object> parsed = objectMapper.readValue(result, MAP_TYPE);
            Object pending = parsed.get("pending");
            if (pending instanceof Map<?, ?> map) return new HashMap<>((Map<String, Object>) map);
        } catch (Exception ignored) {}
        return new HashMap<>();
    }

    private String toPendingJson(Map<String, Object> data) {
        try {
            return objectMapper.writeValueAsString(Map.of("pending", data));
        } catch (Exception e) {
            throw new IllegalStateException("Failed to serialize pending data", e);
        }
    }

    private String serializeList(List<String> list) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < list.size(); i++) {
            if (i > 0) sb.append(", ");
            sb.append("\"").append(list.get(i)).append("\"");
        }
        sb.append("]");
        return sb.toString();
    }
}

File: src/main/kotlin/com/example/dpop/orchestrator/api/v1/identification/fsc/FscIdentificationService.kt
package com.example.dpop.orchestrator.api.v1.identification.fsc

import com.example.dpop.account.AccountProfile
import com.example.dpop.account.AccountService
import com.example.dpop.ext_stammdaten.ExtStammdatenService
import com.example.dpop.id_fsc.IdFscService
import com.example.dpop.orchestrator.account.AccountBindingKeyMappingService
import com.example.dpop.orchestrator.api.v1.AttemptPendingStore
import com.example.dpop.orchestrator.api.v1.OrchestratorException
import com.example.dpop.orchestrator.api.v1.OrchestratorResponse
import com.example.dpop.orchestrator.session.AttemptStatus
import com.example.dpop.orchestrator.session.ChannelSession
import com.example.dpop.orchestrator.session.OrchestratorAttempt
import com.example.dpop.orchestrator.session.ProcessPurpose
import com.example.dpop.orchestrator.session.ProcessSession
import com.example.dpop.orchestrator.session.RegistrationProcessSession
import com.example.dpop.orchestrator.session.SessionManagementService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Duration
import java.util.UUID

@Service
@Transactional
class FscIdentificationService(
    private val sessionManagementService: SessionManagementService,
    private val extStammdatenService: ExtStammdatenService,
    private val idFscService: IdFscService,
    private val accountService: AccountService,
    private val accountBindingKeyMappingService: AccountBindingKeyMappingService,
    private val pendingStore: AttemptPendingStore
) {

    fun startIdentification(
        channelSessionId: UUID,
        bindingKeyRef: String,
        data: Map<String, Any>?
    ): OrchestratorResponse {
        val channelSession = getChannelSession(bindingKeyRef, channelSessionId)
        val channelId = channelSession.channelSessionId
            ?: throw IllegalStateException("Channel session has no id")

        val processSession = sessionManagementService
            .getLatestProcessSessionByChannel(channelId, ProcessPurpose.REGISTRATION)
            .orElseThrow { IllegalArgumentException("Process session not found") }
        @Suppress("UNCHECKED_CAST")
        val registrationSession = processSession as RegistrationProcessSession
        registrationSession.selectedIdentificationMethod = "fsc"
        sessionManagementService.updateProcessSession(registrationSession)

        val processSessionId = registrationSession.processSessionId
            ?: throw IllegalStateException("Process session has no id")

        val attempt = sessionManagementService.createIdentificationAttempt(
            processSessionId, "fsc", "input", Duration.ofMinutes(10)
        )

        val pending = FscPending.empty().merge(data)
        return advance(attempt, pending, bindingKeyRef, channelSession, registrationSession)
    }

    fun submitIdentificationData(
        attemptId: UUID,
        bindingKeyRef: String,
        patchData: Map<String, Any>?
    ): OrchestratorResponse {
        val attempt = sessionManagementService.getAttemptById(attemptId)
            .orElseThrow { OrchestratorException.notFound("Attempt not found") }

        if (attempt.status == AttemptStatus.VERIFIED) {
            throw OrchestratorException.conflict("Attempt is already verified")
        }

        val stored = pendingStore.load(attempt, FscPending::class.java)
        val pending = (stored ?: FscPending.empty()).merge(patchData)

        val channelSession = sessionManagementService.getChannelSessionByBindingKeyRef(bindingKeyRef)
            .orElseThrow { OrchestratorException.forbidden("Channel session not found") }
        val processSessionId = attempt.processSessionId
            ?: throw IllegalStateException("Attempt has no process session id")
        val processSession = sessionManagementService
            .getProcessSessionById(processSessionId)
            .orElseThrow { IllegalArgumentException("Process session not found") }

        return advance(attempt, pending, bindingKeyRef, channelSession, processSession)
    }

    fun getIdentificationStatus(attemptId: UUID, bindingKeyRef: String): OrchestratorResponse {
        val attempt = sessionManagementService.getAttemptById(attemptId)
            .orElseThrow { OrchestratorException.notFound("Attempt not found") }

        val result: Any? = if (attempt.status == AttemptStatus.VERIFIED) attempt.result else null
        return OrchestratorResponse(null, null,
            OrchestratorResponse.AttemptState(
                attempt.attemptId, "identification",
                attempt.status.toString(), null, result),
            OrchestratorResponse.NextRouting(attempt.nextContext, attempt.nextStep)
        )
    }

    private fun advance(
        attempt: OrchestratorAttempt,
        pending: FscPending,
        bindingKeyRef: String,
        channelSession: ChannelSession,
        processSession: ProcessSession
    ): OrchestratorResponse = when (val step = nextStep(pending)) {
        is FscStep.NeedInput -> {
            val missing = step.missingFields
            val attemptId = attempt.attemptId ?: throw IllegalStateException("Attempt has no id")
            val channelId = channelSession.channelSessionId ?: throw IllegalStateException("Channel has no id")
            pendingStore.save(attempt, pending)
            attempt.status = AttemptStatus.INPUT_REQUIRED
            saveMissingFields(attempt, missing)
            sessionManagementService.updateAttempt(attempt)
            OrchestratorResponse(
                channelId,
                OrchestratorResponse.ProcessState("REGISTRATION", "ACTIVE", null, null),
                OrchestratorResponse.AttemptState(attemptId, "identification",
                    "INPUT_REQUIRED", missing, null),
                OrchestratorResponse.NextRouting("fsc", "input")
            )
        }
        is FscStep.Verify -> {
            verifyFsc(attempt, step.kvnr, step.fsc, bindingKeyRef, channelSession, processSession)
        }
    }

    private fun nextStep(p: FscPending): FscStep {
        val missing = p.missingFields()
        return if (missing.isEmpty()) FscStep.Verify(p.kvnr, p.fsc) else FscStep.NeedInput(missing)
    }

    private fun verifyFsc(
        attempt: OrchestratorAttempt,
        kvnr: String?,
        fsc: String?,
        bindingKeyRef: String,
        channelSession: ChannelSession,
        processSession: ProcessSession
    ): OrchestratorResponse {
        val personId = extStammdatenService.findPersonIdByKvnr(kvnr!!).orElse(null)
        if (personId == null || !idFscService.validateFsc(personId, fsc!!)) {
            val attemptId = attempt.attemptId ?: throw IllegalStateException("Attempt has no id")
            attempt.status = AttemptStatus.FAILED
            attempt.result = "{}"
            sessionManagementService.updateAttempt(attempt)
            throw OrchestratorException.verificationFailed("FSC validation failed")
        }

        attempt.status = AttemptStatus.VERIFIED
        attempt.result = "{ \"identified\": true, \"personId\": $personId }"
        sessionManagementService.updateAttempt(attempt)

        processSession.personId = personId
        sessionManagementService.updateProcessSession(processSession)

        val processSessionId = processSession.processSessionId
            ?: throw IllegalStateException("Process session has no id")
        val accountId = channelSession.accountId
            ?: throw IllegalStateException("Channel has no account id")
        val attemptId = attempt.attemptId ?: throw IllegalStateException("Attempt has no id")
        val channelId = channelSession.channelSessionId ?: throw IllegalStateException("Channel has no id")

        val account: AccountProfile = accountService.identifyAccount(
            personId, "fsc", "HIGH", processSessionId,
            java.util.HashMap<String, Any>().apply { put("kvnr", kvnr!!) } as java.util.Map<String, Any>
        )
        sessionManagementService.setAccountId(channelId, account.accountId!!)
        accountBindingKeyMappingService.mapBindingKeyToAccount(bindingKeyRef, account.accountId!!)

        val hasAuth = account.activeAuthenticationMethods.isNotEmpty()
        val next = if (hasAuth)
            OrchestratorResponse.NextRouting("authentication", "selectMethod",
                account.activeAuthenticationMethods, null, account.accountId!!, personId)
        else
            OrchestratorResponse.NextRouting("enrollment", "selectMethod",
                listOf("sms"), null, account.accountId!!, personId)

        return OrchestratorResponse(
            channelId,
            OrchestratorResponse.ProcessState("REGISTRATION", "ACTIVE", personId, account.accountId!!),
            OrchestratorResponse.AttemptState(attemptId, "identification", "VERIFIED",
                null, java.util.Map.of("identified", true, "personId", personId)),
            next
        )
    }

    private fun getChannelSession(bindingKeyRef: String, channelSessionId: UUID?): ChannelSession {
        val cs = sessionManagementService.getChannelSessionByBindingKeyRef(bindingKeyRef)
            .orElseThrow { OrchestratorException.forbidden("Channel session not found") }
        if (channelSessionId != null && cs.channelSessionId != channelSessionId) {
            throw OrchestratorException.forbidden("Channel session mismatch")
        }
        return cs
    }

    private fun saveMissingFields(attempt: OrchestratorAttempt, missing: List<String>) {
        attempt.missingFields = missing.joinToString(", ", "[", "]") { "\"$it\"" }
    }
}
